### DISCUSSION TITLE
LSP4IJ 0.4.0 Release

### DISCUSSION THREAD

**Discussion by angelozerr** [redhat-developer/lsp4ij#452] [Announcements]
Hi,

we are about to release LSP4IJ 0.4.0. Highlights of this release include:
 * Support for workspace/symbol
 * several bug fixes (for completion, syntax coloration in hover, etc)
 
 More details in the <a href="https://github.com/redhat-developer/lsp4ij/milestone/4?closed=1">changelog</a>.
 
 Please give a try to the [latest nightly build](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions/nightly).

 Unless you report a blocking issue, we'll proceed with the release next Monday 08/19/2024.
 
 cc @CppCXY @InSyncWithFoo @seachicken @KUGDev @AlexWeinstein92 @aparnamichael @MituuZ @turkeylurkey @mrglavas @koxudaxi @FalsePattern

**Comment by fbricon** [comment]
We'll actually be releasing on Monday 08/19/2024.

**Comment by mrglavas** [comment]
@angelozerr @fbricon The most recent nightly build I see in the Marketplace is `0.3.1-20240813-013117`. Can I assume this is 0.4.0 for the purpose of testing?

![image](https://github.com/user-attachments/assets/56bb1006-d0ff-4b70-b81a-4c7a453283ab)

**Reply by angelozerr** [reply]
Indeed I changed the version to 0.4.0 in my last push but [latest nightly build](https://plugins.jetbrains.com/plugin/23257-lsp4ij/versions/nightly) has not be rebuilt.

Once https://github.com/redhat-developer/lsp4ij/actions/runs/10374504784 action will be finished, 0.4.0 should be available.

**Reply by angelozerr** [reply]
It should be good now 
![image](https://github.com/user-attachments/assets/dfc6513d-f810-4a54-9a06-d0b0cc7cdbd4)

**Comment by mrglavas** [comment]
When I upgraded to the latest LSP4iJ with our current development stream for Liberty Tools I started seeing this pop-up:

![image](https://github.com/user-attachments/assets/ac0eee0a-b66d-4342-9c04-b68eee6cb3a5)

and the following is appearing in trace for MicroProfile:

```
[Trace - 15:18:17] Received response 'workspace/symbol - (23)' in 2ms.
No result returned.
Error: {
  "code": -32603,
  "message": "Internal error.",
  "data": "java.util.concurrent.CompletionException: java.lang.NullPointerException: Cannot invoke \"java.util.List.stream()\" because the return value of \"java.util.concurrent.CompletableFuture.getNow(Object)\" is null\r\n\tat java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:315)\r\n\tat java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:320)\r\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)\r\n\tat java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)\r\n\tat java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2147)\r\n\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:212)\r\n\tat org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)\r\n\tat org.eclipse.lsp4mp.ls.commons.ParentProcessWatcher.lambda$apply$0(ParentProcessWatcher.java:148)\r\n\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)\r\n\tat org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)\r\n\tat org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)\r\n\tat java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)\r\n\tat java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)\r\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)\r\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)\r\n\tat java.base/java.lang.Thread.run(Thread.java:840)\r\nCaused by: java.lang.NullPointerException: Cannot invoke \"java.util.List.stream()\" because the return value of \"java.util.concurrent.CompletableFuture.getNow(Object)\" is null\r\n\tat org.eclipse.lsp4mp.ls.MicroProfileWorkspaceService.lambda$symbol$3(MicroProfileWorkspaceService.java:105)\r\n\tat java.base/java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:273)\r\n\tat java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)\r\n\tat java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)\r\n\tat java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)\r\n\tat java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)\r\n\tat java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)\r\n\tat java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)\r\n\tat org.eclipse.lsp4mp.ls.MicroProfileWorkspaceService.lambda$symbol$4(MicroProfileWorkspaceService.java:107)\r\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)\r\n\t... 13 more\r\n"
}
```

Guessing this is related to the new function in 0.4.0 based on the changelog descriptions.

**Reply by angelozerr** [reply]
You have this error because you have not implemented jaxrs workspace symbol like we did in IJ Quarkus.

I agree that the error is ugly, it should be fixed on mp ls side to avoid having an npe and return an empty array

**Reply by mrglavas** [reply]
We are closing in on our first release of Liberty Tools which adopts LSP4iJ. We were targeting 0.3.0 prior to today. I think this is an example of a change to LSP4iJ that might have had an impact on our users if we had already released. We would not have seen this with our testing of LSP4iJ 0.3.0 but users would experience this error if they did an install after LSP4iJ 0.4.0 is released (since that would become the latest version of LSP4iJ in the JetBrains Marketplace).

**Reply by angelozerr** [reply]
> We are closing in on our first release of Liberty Tools which adopts LSP4iJ. We were targeting 0.3.0 prior to today. I think this is an example of a change to LSP4iJ that might have had an impact on our users if we had already released. We would not have seen this with our testing of LSP4iJ 0.3.0 but users would experience this error if they did an install after LSP4iJ 0.4.0 is released (since that would become the latest version of LSP4iJ in the JetBrains Marketplace).

I understand the problem, but here, the problem comes from the MicroProfile language server which have a bug. LSP4IJ which supports workspace symbol highlights this bug.

Since we have published LSP4IJ we have the luck to have several adaptors https://github.com/redhat-developer/lsp4ij?tab=readme-ov-file#who-is-using-lsp4ij so we need to fix bugs quickly and provide new features for them. It is one reason why we create little release like 0.4.0.

I suggest really that you follow our PRs done in LSP4IJ and IJ Quarkus to avoid discovering before the release new features like you discover now. In my case, I'm following your Liberty IntelliJ repository to see if you have some trouble with LSP4IJ or see if you fix some bugs from copied code.

**Reply by mrglavas** [reply]
We are already actively monitoring PRs in LSP4iJ and Quarkus Tools. This is part of the focus (https://github.com/OpenLiberty/liberty-tools-intellij/issues/830) of our upcoming release, automating notifications to our team and continuously integrating changes to LSP4iJ (including open PRs) to test that the LTI plugin continues working. I know you weren't available that day but in our last discussion with @fbricon we were made aware of this feature as something that would be coming but also not to expect it in the near future. The timing of this release is a surprise. We thought there would be more time to evaluate these changes before they're released to the Marketplace. I think this would be the fourth release in three months. Is this the actual release cadence we should be expecting for LSP4iJ?

Hopefully in the near future we will release the first version of Liberty Tools which adopts LSP4iJ. Of course development will continue here and we'll be following it, but we can only follow PRs when they exist and we can only see new changes when they're available. Our concern is that future changes do not have an impact on our existing releases which is why I'm providing feedback on how this change would have affected our users had we already released our plug-in in the Marketplace. Could this new feature have been introduced in such a way that it doesn't affect an adopter of LSP4iJ unless they opt-in? I think this would have mitigated the risk to released plug-ins which use LSP4iJ.

**Reply by angelozerr** [reply]
As I tried to explain you, the bug comes from MicroProfile LS. I will try do a fix it before the release of LSP4MP. 

If I understand, you wait for the LSP4MP release to do a release of Open Liberty, I will do the fix beforethe release of LSP4MP in order to avoid the error you have.

**Reply by mrglavas** [reply]
I understand where the bug comes from but also that the bug was not exposed until this change. In my years of maintaining other components (e.g. Xerces) I have had to be very careful about introducing changes that might break applications even if the change in isolation seems to be correct. It's the overall impact that matters.

Fortunately we have not released yet and were already planning to wait for LSP4MP since our development stream currently depends on a snapshot of 0.12.0. I believe we'd want a fix there before we move forward. Thanks for taking a look at this.

**Reply by angelozerr** [reply]
@mrglavas I created the PR for MicroProfile LS https://github.com/eclipse/lsp4mp/pull/457

**Reply by fbricon** [reply]
Sorry this is causing some troubles, I agree this is unfortunate it happened in the first place, but the upside is this announcement fulfilled its objectives, i.e. a problem was caught before the release and we could address it (https://github.com/eclipse/lsp4mp/pull/457) before any harm was done.

As I mentioned before, our focus is now on resuming Quarkus maintenance, to keep up with VS Code. Part of that is getting Workspace Symbols working in IJ. That and other bug fixes in LSP4IJ were required for us to move forward with our planned release next week (Thursday 22/08/24), which will include LSP4MP*, Quarkus LS, vscode-microprofile, vscode-quarkus and intellij-quarkus. 

Unfortunately there is currently no way for adopters to selectively opt-in into new features, but we'd like adopters to be able to override (thus disable) selected features. This will require non-trivial changes, so if you want to help providing API customization, please take a look at https://github.com/redhat-developer/lsp4ij/issues?q=is%3Aissue+is%3Aopen+label%3Aapi-customization.

BTW, another needed (safe) [fix](https://github.com/redhat-developer/lsp4ij/pull/456) was merged into LSP4IJ, if you want to retest another nightly build.

\* you should be able to test an LSP4MP build this Friday, once those [last 2 LSP4MP PRs](https://github.com/eclipse/lsp4mp/milestone/13) are merged.

**Reply by mrglavas** [reply]
Thanks. Planning to test LSP4MP once a build is available and will give LSP4iJ 0.4.0 another spin.

I'm still trying to wrap my head around this whole ecosystem. This would all be less of an issue if JetBrains allowed plugins to control the versions of their dependencies, at least when they're initially installed. The challenge with us being able to utilize API customization is how could we apply that in advance of an unknown future feature. I can imagine how that could be helpful for a future release of our plugin but is there some way for it to be used with an existing release in the Marketplace? With our focus on staying current with IntellIJ major releases we may be releasing much less frequently than LSP4iJ. If LSP4iJ has a release or multiple releases in the time between Liberty Tools releases we want to make sure our current release continues to work with the current version of LSP4iJ that our users would be installing as a dependency when they install Liberty Tools. We have our build automation ready to test that scenario (https://github.com/OpenLiberty/liberty-tools-intellij/issues/815) and would provide feedback should we find something like we did here.

**Reply by angelozerr** [reply]
@mrglavas I did the fix and support System/Env var (see https://github.com/eclipse/lsp4mp/issues/448) and the 0.12.0 SNAPSHOT should be available https://download.eclipse.org/justj/?file=lsp4mp/snapshots/0.12.0/repository/plugins, please try it and give us feedback.

**Reply by mrglavas** [reply]
With the new snapshot I'm not seeing the `workspace/symbol` error anymore. The trace looks normal.

![image](https://github.com/user-attachments/assets/af5e6aae-3610-4052-9588-99ff0b03f736)

**Reply by mrglavas** [reply]
Regarding https://github.com/eclipse/lsp4mp/issues/448 is this what I should be expecting to see? I had to hover to see the resolved value.

![image](https://github.com/user-attachments/assets/9ff86fcd-f289-4122-8642-479ecd6712e2)

**Reply by angelozerr** [reply]
LSP4IJ support LSP InlayHint, here the result with IJ Quarkus:

![image](https://github.com/user-attachments/assets/6f0aba36-9f3c-45f7-8b7c-6bc4a1815b17)

I suspect that you have not activate the Inlay Hint for MicroProfile LS. I suggest you see our IJ Quarkus.

**Reply by mrglavas** [reply]
> I suspect that you have not activate the Inlay Hint for MicroProfile LS. I suggest you see our IJ Quarkus.

@turkeylurkey @aparnamichael We should investigate this. Not sure if we missed integrating something or if this is brand new, but think we should check this out when we synch up LSP4MP4iJ from IJ Quarkus. See https://github.com/OpenLiberty/liberty-tools-intellij/issues/918.

**Comment by mrglavas** [comment]
This is more a question about user experience... I noticed that there's a `"Semantic Tokens Inspector"` tab showing up in the IDE. Not sure if this is new for 0.4.0 but I assume this is coming from LSP4iJ. I had to look up what a "Semantic Token" is and still not sure after reading that what these checkbox configuration options do. I do wonder if the average user seeing this tab would understand what it's for and how to use it. Perhaps it needs more explanation somewhere, maybe in the blank space on the right side of the panel?

![image](https://github.com/user-attachments/assets/46fa9ace-00ce-4d17-904e-ca7f10aba284)

Can this be hidden if we determine this doesn't apply to anything supported by our plugin?

**Reply by angelozerr** [reply]
It is afeature from 0.3.0, see https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserGuide.md#semantic-tokens-support

This tab should be hidden by default, no?

**Reply by InSyncWithFoo** [reply]
> This tab should be hidden by default, no?

I don't think so. A bug, maybe?

**Reply by mrglavas** [reply]
> It is afeature from 0.3.0, see https://github.com/redhat-developer/lsp4ij/blob/main/docs/UserGuide.md#semantic-tokens-support
> 
> This tab should be hidden by default, no?

We just moved to 0.3.0, which was just an update to our `build.gradle` to use 0.3.0 from the Marketplace: https://github.com/OpenLiberty/liberty-tools-intellij/commit/8287c40a55a83c8dc0a6d2081fbacd7a8a59a971. In my testing I'm updating the value to the latest nightly build so it appears we're getting this by default.

**Reply by angelozerr** [reply]
> > This tab should be hidden by default, no?
> 
> I don't think so. A bug, maybe?

I did that very quickly, any contribution are welcome to improve it.

**Reply by angelozerr** [reply]
I dont know how to do that (and no time to investigate that), I created the issue https://github.com/redhat-developer/lsp4ij/issues/461 any contribution are welcome!

**Comment by angelozerr** [comment]
I launched the 0.4.0 release , JetBrains is reviewing it.

**Comment by angelozerr** [comment]
LSP4IJ 0.4.0 is available at the JetBrains Marketplace https://plugins.jetbrains.com/plugin/23257-lsp4ij/edit/versions/stable/590299

**Reply by angelozerr** [reply]
@mrglavas please note that we have released yesterday lsp4mp and IJ Quarkus which are now based on LSP4IJ

**Reply by turkeylurkey** [reply]
Thank you, noted.
